### ISSUE TITLE
ADO.NET SqlClient DbConnectionOptions: Column Encryption Setting as Parameter

### ISSUE THREAD

**Issue by phhbr** [dotnet/corefx#18276]
When using a DbContext with a ConnectionString like

```
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=.\\SQLEXPRESS;Database=EFGetStarted.AspNetCore.NewDb;Trusted_Connection=True;ColumnEncryptionSetting=Enabled;MultipleActiveResultSets=true;"
  }
}
```

I get the following error
StackTrace
```
{System.ArgumentException: Keyword not supported: 'columnencryptionsetting'.
   at System.Data.Common.DbConnectionOptions.ParseInternal(Dictionary`2 parsetable, String connectionString, Boolean buildChain, Dictionary`2 synonyms)
   at System.Data.Common.DbConnectionOptions..ctor(String connectionString, Dictionary`2 synonyms)
   at System.Data.SqlClient.SqlConnectionString..ctor(String connectionString)
   at System.Data.SqlClient.SqlConnectionFactory.CreateConnectionOptions(String connectionString, DbConnectionOptions previous)
   at System.Data.ProviderBase.DbConnectionFactory.GetConnectionPoolGroup(DbConnectionPoolKey key, DbConnectionPoolGroupOptions poolOptions, DbConnectionOptions& userConnectionOptions)
   at System.Data.SqlClient.SqlConnection.ConnectionString_Set(DbConnectionPoolKey key)
   at System.Data.SqlClient.SqlConnection.set_ConnectionString(String value)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerConnection.CreateDbConnection()
   at Microsoft.EntityFrameworkCore.Internal.LazyRef`1.get_Value()
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.BeginTransaction(IsolationLevel isolationLevel)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(Tuple`2 parameters)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](Func`2 operation, Func`2 verifySucceeded, TState state)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at xBN.Repository.Repositories.EntityBaseRepository`1.Commit() in C:\!Checkouts\git checkouts\xBN\xBN.Service\Repositories\EntityBaseRepository.cs:line 98
   at xBN.API.Controllers.CompanyController.Get(Int32 id) in C:\!Checkouts\git checkouts\xBN\xBN.API\Controllers\CompanyController.cs:line 37
   at lambda_method(Closure , Object , Object[] )
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionMethodAsync>d__27.MoveNext()}
```

The issue is about adding support for the Column Encryption Setting to the .NET Core ADO.NET SqlClient.

**Comment by jasondaicoder** [comment]
guys, has this been fixed or?

**Comment by corivera** [comment]
CoreFx SqlClient does not currently support column encryption

**Comment by jasondaicoder** [comment]
@corivera, do we have any plans on this feature?

**Comment by corivera** [comment]
We don't have any plans for it at the moment, we (SqlClient team) will have to discuss it.

**Comment by jasondaicoder** [comment]
Thanks for you info. I think it would be highly demanded feature as column encryption is widely used. Hope to see this feature being implemented soon.

**Comment by divega** [comment]
Related https://github.com/aspnet/EntityFramework/issues/9142.

**Comment by LarrySmith-1437** [comment]
The lack of Always Encrypted is a real drag on promoting .net Core where I am, because security is paramount.

**Comment by wizofaus** [comment]
Actually I had reported this separately, so technically my issue is a duplicate. But I've been looking at explicit server-based value encryption (using ENCRYPTBYKEY etc), and it seems there are no .net libraries out there that support this either?

**Comment by aspnerd** [comment]
What is going on with this? We really need this now!

**Comment by danmosemsft** [comment]
@divega @corivera is this in plan? If not is it feasible for community to port?

**Comment by mattd73uk** [comment]
Just starting a new enterprise project, we had decided on .Net Core but are switching to .Net Framework over this issue - I just removed the .Net Core projects from the solution and am retargeting ASP.Net Core at the .Net Framework. It is a shame Core isn't yet mature enough for us, but at the end of the day you've got the resources you've got.

**Comment by JohnKoz** [comment]
@corivera : Similar to @mattd73uk, we've started an enterprise project and just found out always encrypted is not supported in .Net Core. Basically we will have to switch back to the .Net Framework unless we hear otherwise. Is there any indication on when this feature would likely be added?

**Comment by se7vanj** [comment]
We had the same demand during mid of 2017 in one of our enterprise project that uses Always Encrypted DB from a .NET Core API, reached through the Microsoft premier support, but could not find any timelines for this functionality, I don't think its even present in their future milestones. 

Wondering when they will add support for column encryption in CoreFX sql client.

Finally, we had to target the API to .NET Framework and use EF 6.x instead of Core.

**Comment by danmosemsft** [comment]
@divega @corivera could you please share thoughts? Clearly it's a blocking issue for a number of folks here. Is it something community could help port?

**Comment by louislewis2** [comment]
Ouch, this has burned me today horribly. Tomorrow we are deploying to production. Which means that today we start migrating users across from an old system across to a new one.

Then I get hit by this......

**Comment by ld0614** [comment]
While I was looking forward to trying .NET Core out for a new project the lack of support for column encryption and the lack of support for AccessToken means I will have to build the project on .Net Framework instead.

**Comment by David-Engel** [comment]
Always Encrypted is definitely still on our radar. We don't have a timeline for it at the moment, though. There are a few items that we believe are higher on the list. We are open to prioritization changes, though (vote up comments on this issue).

**Comment by rudiv** [comment]
I've commented on another thread regarding this already - but with your recent push on GDPR it shocks me that this isn't available. It's fairly critical to stop people from rolling their own methods to encrypt data in columns (which as we know from experience never ends well).

Being forced to use .NET Framework for such "basic" (from an outside perspective) functionality isn't really acceptable, when there's such a huge push to use .NET Core.

**Comment by aspnerd** [comment]
@David-Engel We cannot use the latest SQL and .net Core due to this. We had hours and hours of rework which had to be done because of this and we aren't the only ones. This should have already been done imo.

**Comment by SariDev** [comment]
I hope to see this feature in the next version of System.Data.SqlClient.
Support for Always Encrypted is required for enterprise applications that rely on SQL Server and it's security features.
Releasing this together with .NET Core 2.2 later this year should be high priority.

**Comment by filippostefanello** [comment]
I share your hope as I need a solution to comply the GDPR's rules.

**Comment by armordba** [comment]
With all the data security push from GDPR and others, we really need this fix ASAP

**Comment by RJoseSilva** [comment]
we really need this fix ASAP. This feature will be release with .Net Core 2.2 ?

**Comment by tuath** [comment]
Any news on this ? would be nice to not have to use .net framework for bigger projects...

**Comment by mjtpena** [comment]
Wow. Basically my whole Cryptography layer can't be upgraded to .NET Core because of this.

**Comment by houseofcat** [comment]
Would very much like this prioritized.

Understand you are busy, do you guys want help - this isn't really an optional feature.

Rolling custom encryption is going to lead to insecurity, which is really bad PR for NetCore. It's also been lower priority than features released and then re-released under NetCore 1.0 and then 2.0.

**Comment by David-Engel** [comment]
Always Encrypted is still very much a priority for the SqlClient team and we are hearing all the comments. Unfortunately, we still don't have a concrete timeline we can share yet.

**Comment by aspnerd** [comment]
We are going 3 versions deep in SQL Server without support. We really want to use .net core, but we can't because of this we have to stay full framework. When will it be priority enough to get it finished?

**Comment by houseofcat** [comment]
@David-Engel can we bribe you with cookies and beer?

**Comment by David-Engel** [comment]
@houseofcat I do like a good chocolate chip cookie. 😄 But some things are beyond my control. I wish I could share more. There are always other factors in play but we are definitely are shooting for AE support in .NET Core as soon as we can.

**Comment by gulbanana** [comment]
I have a .NET Core application currently using SQL Server. In a future release, data is being added to the system which needs column encryption due to its nature - we’re going to have to port it to .NET Framework due to this missing capability.

**Comment by mentallabyrinth** [comment]
Was always told by pundits that .NET had best in class security which was the governing reason for selecting the platform. Surprised to find out that as of yet ".NET Core" doesn't support the Always Encrypted feature. Neither our team or the consultants we hired were aware of this 😆Learning experience, tenfold. 

Anyway, throwing our hat into the ring to support fast-tracking this feature, it would be beneficial to have. Because conversion is out of the question, but we'll figure out something else in the interim. 

And it's worth saying that despite this, thanks for the excellent product so far! 😀

**Comment by tomRedox** [comment]
Adding our name to the list of people not implementing new projects in .NET Core because of the lack of AE support.

**Comment by jxvalenz** [comment]
Work for a large Fortune 10 company that is also putting the brakes on many large projects due to this. Going with .NET Framework is not an option. We would really appreciate it if at least you could give us hope that this feature is coming in the near future, at least in beta form, so our management doesn't throw .NET Core out the window and go for a Java solution. We have already invested in .NET Core and was about to become our corporate standard. Please give us a date range at least.

**Comment by aspnerd** [comment]
I wonder if the holdup is the fact that the encryption is proprietary and they can't open source it. If that is the cast then they should just create an add on nuget package for .net core so we can use it!

**Comment by lsuarez5280** [comment]
@aspnerd that definitely isn't the problem. The existing SqlClient always encrypted support code is available right now in the Framework GitHub. There's nothing proprietary about that implementation.

**Comment by David-Engel** [comment]
@jxvalenz It's coming. Someone is actively working on the code. I don't have a timeline I can publicly share right now, though.

**Comment by houseofcat** [comment]
@David-Engel thank you for the update and I understand you can't give specifics, however, what we really could us are some abstracts. Will we have a preview/release this year (2018), or in v2.2, >v2.2, 3.0. etc.

Can you let us know if there is intention to release it this year? 

This isn't asking for a concrete delivery date, this is so we can plan our future NetCore empires and we have to prototype and sell the ideas we are creating - and the technology that powers are ideas - to Project Managers who by-and-large eat q-tips.

**Comment by kevbry** [comment]
Seconding @houseofcat . With full framework support going away in asp.net core 3.0 we need to start making plans.

**Comment by gulbanana** [comment]
This has definitely become more urgent with the news that ASP.NET Core will no longer run on netfx - the plan of porting back to netfx which we had is no longer feasible!

**Comment by aspnerd** [comment]
@divega I don't this is related to our issue above.

**Comment by divega** [comment]
@aspnerd they are related in the sense that they are both potentially competing priorities for the SqlClient team to tackle in the near future. You can follow the link to see in what context I mentioned it.

**Comment by mattd73uk** [comment]
Hi, as per my comment above I switched our enterprise project from Core to Framework over this, definitely the right decision, SQL Server 2017 with ASP.Net Core targeting the .Net Framework with EF is a great combination of secure & easy to maintain tech.

But taking a step back, my concern is how is it possible that the latest Microsoft technology in one area (.Net Core) has spent two years not supporting/interop with such a great feature in the current MS tech in another area (SQL Server 2016, 2017, 2019).  

I don't mind using the .Net Framework - it's fantastic & served me well.  But from its announcements the MS megalith (that's a good thing - I and my clients have relied on it for 20 years) *wants* me/my clients to use .Net Core going forward, but it hasn't able to put the resources in play to make what it wants happen.  

@divega does that make sense as a concern?  Yes, you are able to cite an issue that has received more upvotes, but why is there a need to pit two so important/requested issues against each other?  Why has MS been unable to resource either for 2 years to support a flagship product like SQL Server?

**Comment by achancecwg** [comment]
We are developing a LOB application using UWP and just hit a roadblock because of the lack of support for this feature. I think that these types of feature are critical for developers to be able to adopt UWP and .NET Core/ .NET Standard, especially for Enterprise applications. 

Would there need to be a separate issue opened for UWP support?

Please consider implementing this feature soon!

**Comment by vickyharp** [comment]
I echo what @David-Engel says about not being able to share a concrete timeframe, but I can tell you that the feature is under active development with high priority for the SqlClient team. This capability is important to our users and it is important for cross platform parity in Microsoft's own data tools.

For directional purposes: it is currently planned for the v3.0 timeframe and we will announce more soon as we become more confident in the timing.

**Comment by houseofcat** [comment]
My body is ready!

**Comment by ToonVanhoutte** [comment]
Highly demanded feature!!!  Incredible that there's still no support for this in .NET Core, after two years!

**Comment by divega** [comment]
As mentioned by @vickyharp previously, it is now planned for 3.0. I have updated the milestone assignment to reflect this.

**Comment by mmacagno** [comment]
I see .NET Core 3.0 previews are coming out now. Is this still in scope for the release?

**Comment by vickyharp** [comment]
@mmacagno Yes

**Comment by davidsaldana** [comment]
Man, hopefully AE support gets rolled out soon. We need this big time...

**Comment by Runamok81** [comment]
Adding in my vote for this as well.  Our project had to switch from .NET Core to .NET Framework over this exact limitation.

**Comment by TrayceJ** [comment]
Will the 3.0 release support the new Always Encrypted enclaves coming out in SQL 2019?  Will there need to be anything done on the .net Core side?   I don't think there would be, but if there is -- will we need to wait another 3 years for that support?   When will there be previews for AE support that we can start testing?

**Comment by gbelzile** [comment]
A workaround is to use ODBC driver to access AE data. It works with .net core 2.x.
https://docs.microsoft.com/en-us/sql/connect/odbc/using-always-encrypted-with-the-odbc-driver?view=sql-server-2017. In my project I have 2 connection strings, one SqlServer for non encrypted data and one ODBC for everything that needs to be decrypted. It is a pain but it works well.

**Comment by LarrySmith-1437** [comment]
> 
> 
> A workaround is to use ODBC driver to access AE data. It works with .net core 2.x.
> https://docs.microsoft.com/en-us/sql/connect/odbc/using-always-encrypted-with-the-odbc-driver?view=sql-server-2017. In my project I have 2 connection strings, one SqlServer for non encrypted data and one ODBC for everything that needs to be decrypted. It is a pain but it works well.

That looks crazy compliated

**Comment by marcoplaut** [comment]
Is it already available on .net core 3.0 preview ? any preview preview release targeted for that feature ?

**Comment by Tim-Bijnens** [comment]
I guess this is still not in any preview? This issue is making porting our website to .Net Core a real pain.

**Comment by divega** [comment]
Everyone,

A new out-of-band version of SqlClient was released in preview today on [NuGet](https://www.nuget.org/packages/Microsoft.Data.SqlClient/). It is a new package called Microsoft.Data.SqlClient, and it includes support for Always Encrypted on .NET Core and other enhancements like UTF-8 and data classification support on both .NET Core and .NET Framework. 

Initial details about the new package are included in the new .NET Core 3.0 Preview 5 announcement at https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0-preview-5/. 

In coming weeks, we will move issue tracking to the new repository at https://github.com/dotnet/sqlclient, where the open source codebase for the new provider will be hosted. In the meantime, issues and PRs that apply to System.Data.SqlClient are still accepted here. 

We would like to encourage you to try the new package and give us feedback, especially if you were blocked by Always Encrypted not being available on .NET Core. 

cc @David-Engel @vickyharp

**Comment by sebader** [comment]
@divega thanks for the awesome news!

I just wanted to try to migrate a sample project from .NET Framework 4.7 to .NET Core 2.2 that is using Always Encrypted on a Azure SQL Database and Azure KeyVault. It was originally built from [this example](https://docs.microsoft.com/de-de/azure/sql-database/sql-database-always-encrypted-azure-key-vault#register-the-azure-key-vault-provider).
But I'm failing so far as `SqlColumnEncryptionKeyStoreProvider` is not available in Microsoft.Data.SqlClient and neither is `SqlConnection.RegisterColumnEncryptionKeyStoreProviders()`

Could you maybe please post a complete example using Always Encrypted with KeyVault in the new Microsoft.Data.SqlClient package?

**Comment by Tim-Bijnens** [comment]
I still have the same issue with Column Encryption Setting in the connection string in my .Net Core app that uses EF.
I use all the previews that were released the 6th of may 2019.
I have the following related NuGet packages installed:

- Microsoft.EntityFrameworkCore v3.0.0-preview5.19227.1
- Microsoft.EntityFrameworkCore.SqlServer v3.0.0-preview5.19227.1
- System.Data.SqlClient v4.7.0-preview5.19224.8
- Microsoft.Data.SqlClient v1.0.19123.2-Preview

 But the app (EntitityFrameworkCore) seems to still have a dependency System.Data.SQLClient instead of the new Microsoft.Data.SqlClient this is the Exception :

ArgumentException: Keyword not supported: 'column encryption setting'.
System.Data.SqlClient.SqlConnectionStringBuilder.GetIndex(string keyword)
System.Data.SqlClient.SqlConnectionStringBuilder.set_Item(string keyword, object value)
System.Data.Common.DbConnectionStringBuilder.set_ConnectionString(string value)
Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerConnection.get_IsMultipleActiveResultSetsEnabled()
Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Microsoft.EntityFrameworkCore.Storage.IRelationalConnection.UnregisterBufferable(IBufferable bufferable)
Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable<T>+AsyncEnumerator.Dispose()
System.Linq.AsyncEnumerable+SelectEnumerableAsyncIterator<TSource, TResult>.Dispose() in Select.cs
Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider+ExceptionInterceptor<T>+EnumeratorExceptionInterceptor.Dispose()
System.Linq.AsyncEnumerable.Aggregate_<TSource, TAccumulate, TResult>(IAsyncEnumerable<TSource> source, TAccumulate seed, Func<TAccumulate, TSource, TAccumulate> accumulator, Func<TAccumulate, TResult> resultSelector, CancellationToken cancellationToken) in Aggregate.cs
ClementinePortal.Core.Pages.Companies.IndexModel.OnGetAsync() in Index.cshtml.cs
+
            Companies = await _context.Companies.ToListAsync();

 is there something I can do to alleviate the issue?

**Comment by sebader** [comment]
@Tim-Bijnens your issue as actually a bit different, as the reason for yours is that EF Core does not yet use the new package. The exception is thus thrown by the old *System.Data.SqlClient*
There is already an issue for the update here: https://github.com/aspnet/EntityFrameworkCore/issues/15636

**Comment by divega** [comment]
> But I'm failing so far as SqlColumnEncryptionKeyStoreProvider is not available in Microsoft.Data.SqlClient and neither is SqlConnection.RegisterColumnEncryptionKeyStoreProviders()
> Could you maybe please post a complete example using Always Encrypted with KeyVault in the new Microsoft.Data.SqlClient package?

@David-Engel would you be able to answer this? I think it is the second time the idea of having a minimal sample comes up. 

> I still have the same issue with Column Encryption Setting in the connection string in my .Net Core app that uses EF.,,

@Tim-Bijnens good point. Unfortunately the current preview of EF Core still uses System.Data.SqlClient. We are moving to Microsoft.Data.SqlClient in an upcoming preview. You can follow https://github.com/aspnet/EntityFrameworkCore/issues/15636 to know when the change will make it to nightly builds.

**Comment by David-Engel** [comment]
@divega It appears that when the AE port to Core was done, the Core ref classes were not updated with the new APIs. Tests were/are run against the non-ref implementation so we didn't catch it. We'll get an update to the package out ASAP.

**Comment by David-Engel** [comment]
We pushed an update to the Microsoft.Data.SqlClient 1.0 preview. It should resolve the issues noted around Always Encrypted.

@sebader - I want to note, though, custom key store providers like the Azure Key Vault provider need to be updated to work with Microsoft.Data.SqlClient. Only the built-in providers will work until that happens (hopefully soon!). So the example you were following using Azure Key Vault still won't work. But this example, which uses the built-in Windows Certificate Store provider, should work as expected after changing from System.Data.SqlClient to Microsoft.Data.SqlClient:
https://docs.microsoft.com/en-us/azure/sql-database/sql-database-always-encrypted

@divega

**Comment by sebader** [comment]
thanks for the update @David-Engel ! Too bad though about KeyVault missing yet. I'm using Azure Functions so Windows Cert store is no good for me ;-)
Where could we track the progress of their update?

**Comment by David-Engel** [comment]
@sebader I'm not sure. We've pinged the team who owns it. I don't think that project is public.

**Comment by divega** [comment]
As recently [announced in the .NET Blog](https://devblogs.microsoft.com/dotnet/introducing-the-new-microsoftdatasqlclient/), focus on new SqlClient features an improvements is moving to the new [Microsoft.Data.SqlClient package](https://www.nuget.org/packages/Microsoft.Data.SqlClient/). For this reason, we are moving this issue to the new repo at https://github.com/dotnet/SqlClient. We will still use https://github.com/dotnet/corefx to track issues on other providers like System.Data.Odbc and System.Data.OleDB, and general ADO.NET and .NET data access issues.
